src/js/ProviderManager.js

```diff
@@ -0,0 +1,45 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at http://mozilla.org/MPL/2.0/. */
+/*globals async, DirectoryLinksProvider, Links, PlacesProvider, gUserDatabase, gPinnedLinks, gBlockedLinks*/
+
+/**
+ * Singleton that handles link providers.
+ */
+
+ "use strict";
+
+const ProviderManager = {
+  init: async(function*(gMockFetch) {
```

Comment:
Hmm. the tooling for JS sucks. However, depending on how JS handles function declarations, we may be able to keep the code similar, without resorting to adding the mock in the init and without using an external service, if it's hard to get the wanted response.

For instance, in the test, we can monkeypatch fetch in the global scope, and add it back when the test completes.

This assumes tests running thread-safe. Which karma probably does.